grakn-engine/src/main/java/ai/grakn/engine/postprocessing/PostProcessor.java

```diff
@@ -0,0 +1,274 @@
+
+            //Update counts
+            jobs.forEach((key, value) -> {
+                metricRegistry
+                        .histogram(name(PostProcessor.class, "shard-size-increase"))
+                        .update(value);
+                Timer.Context contextSingle = metricRegistry
+                        .timer(name(PostProcessor.class, "execution-single")).time();
+                try {
+                    if (updateShardCounts(redis, keyspace, key, value, shardingThreshold)) {
+                        conceptToShard.add(key);
+                    }
+                } finally {
+                    contextSingle.stop();
+                }
+            });
+
+            //Shard anything which requires sharding
+            conceptToShard.forEach(type -> {
+                Timer.Context contextSharding = metricRegistry.timer("sharding").time();
+                try {
+                    shardConcept(redis, factory, keyspace, type, maxRetry, shardingThreshold);
+                } finally {
+                    contextSharding.stop();
+                }
+            });
+            LOG.debug("Updating instance count successful for {} tasks", jobs.size());
+        } catch(Exception e) {
+            LOG.error("Could not terminate task", e);
+            throw e;
+        }
+    }
+
+    /**
+     * Updates the type counts in redis and checks if sharding is needed.
+     *
+     * @param keyspace The keyspace of the graph which the type comes from
+     * @param conceptId The id of the concept with counts to update
+     * @param value The number of instances which the type has gained/lost
+     * @return true if sharding is needed.
+     */
+    private static boolean updateShardCounts(
+            RedisCountStorage redis, Keyspace keyspace, ConceptId conceptId, long value, long shardingThreshold){
+        long numShards = redis.getCount(RedisCountStorage.getKeyNumShards(keyspace, conceptId));
+        if(numShards == 0) numShards = 1;
+        long numInstances = redis.adjustCount(
+                RedisCountStorage.getKeyNumInstances(keyspace, conceptId), value);
+        return numInstances > shardingThreshold * numShards;
+    }
+
+    /**
+     * Performs the high level sharding operation. This includes:
+     * - Acquiring a lock to ensure only one thing can shard
+     * - Checking if sharding is still needed after having the lock
+     * - Actually sharding
+     * - Incrementing the number of shards on each type
+     *
+     * @param keyspace The database containing the {@link ai.grakn.concept.Type} to shard
+     * @param conceptId The id of the concept to shard
+     */
+    private void shardConcept(RedisCountStorage redis, EngineGraknTxFactory factory,
+                              Keyspace keyspace, ConceptId conceptId, int maxRetry, long shardingThreshold){
+        Lock engineLock = lockProvider.getLock(getLockingKey(keyspace, conceptId));
+        engineLock.lock(); //Try to get the lock
```

Comment:
Is this a local lock? Should we get a distributed lock like in the duplicate method?

Reply:
The implementation is `JedisLockProvider` which is distributed. You know we may be able to get rid of the interface. We only have one implementation of `LockProvider` what do you think?

Reply:
Ok! We can leave the interface